rust/src/tx_builder.rs

```diff
@@ -0,0 +1,164 @@
+use super::*;
+use super::fees;
+use super::utils;
+use std::collections::BTreeSet;
+
+// We need to know how many of each type of witness will be in the transaction so we can calculate the tx fee
+struct MockWitnessSet {
+  vkeys: BTreeSet<AddrKeyHash>,
+  scripts: BTreeSet<ScriptHash>,
+  bootstraps: BTreeSet<String>,
+}
+
+#[wasm_bindgen]
+pub struct TransactionBuilder {
+    inputs: TransactionInputs,
+    outputs: TransactionOutputs,
+    fee: Option<Coin>,
+    ttl: Option<u32>, // absolute slot number
+    certs: Option<Certificates>,
+    withdrawals: Option<Withdrawals>,
+    metadata: Option<TransactionMetadata>,
+    input_types: MockWitnessSet,
+}
+
+// comes from witsVKeyNeeded in the Ledger spec
+fn witness_keys_for_cert(cert_enum: &Certificate, keys: &mut BTreeSet<AddrKeyHash>) {
+    match &cert_enum.0 {
+        CertificateEnum::StakeRegistration(cert) => {
+            keys.insert(cert.stake_credential().to_keyhash().unwrap());
+        },
+        CertificateEnum::StakeDeregistration(cert) => {
+            keys.insert(cert.stake_credential().to_keyhash().unwrap());
+        },
+        CertificateEnum::StakeDelegation(cert) => {
+            keys.insert(cert.stake_credential().to_keyhash().unwrap());
+        },
+        CertificateEnum::PoolRegistration(cert) => {
+            for owner in &cert.pool_params().pool_owners().0 {
+                keys.insert(owner.clone());
+            }
+            keys.insert(
+                AddrKeyHash::from_bytes(cert.pool_params().operator().to_bytes()).unwrap()
```

Comment:
In the 3 places in this function, we can safely convert these to `AddrKeyHash` because both `AddrKeyHash` and the other key hash types all have the same hash size and are generated by the same hashing algorithm.
Therefore, the hash matching => the underlying key matches

The code is a slightly ugly way to represent that though. Not sure if there's is a better way though.
One alternative is to add a generic From/Into between all the KeyHash types in `crypto.rs` (`GenesisDelegateHash`, `PoolKeyHash`, `AddrKeyHash`, `VRFKeyHash`). However, that doesn't really feel less hacky (but maybe more readable)

Reply:
Yes, in terms of size it doesn't matter since scripthash and addrkeyhash are 28 bytes, but the problem is that as-is that unwrap will panic for scripthash staking credentials as `StakingCredential::to_keyhash()` returns `None` for script credentials. I see you're referring to the other hashes, but for `StakingCredential` it's not a matter of ugly, it will just panic using `to_keyhash()`.

We could do a conversion you said and make some Hash28/Hash32 (or whatever you want to call them - just using those since the shelley.cddl uses them) and make this a `BTreeSet<Hash28>` so it doesn't seem like we're reusing AddrKeyHash for hashes of other keys just because it's the same size.